### DISCUSSION TITLE
Announce  supported capabilities

### DISCUSSION THREAD

**Discussion by predragnikolic** [vuejs/language-tools#393] [General]
LSP clients(like Sublime Text), won't send server requests if the server didn't announce support for those requests through capabilities.

Currently the Volar server only announces this capability:
https://github.com/johnsoncodehk/volar/blob/8733e9483389d9567525f0493b56697fe8f2c35a/packages/server/src/index.ts#L34-L36

But it should actually include lots of other capabilities.
Here is an example of what the Vetur server returns:
<details>

Note: I accidentally lowercased all the object keys :)

```json
{
	"capabilities": {
		"foldingrangeprovider": true,
		"codeactionprovider": {
			"codeactionkinds": ["quickfix", "refactor", "refactor.extract", "refactor.inline", "refactor.rewrite", "source", "source.organizeimports"],
			"resolveprovider": true
		},
		"signaturehelpprovider": {
			"triggercharacters": ["("] 
		},
		"colorprovider": true,
		"semantictokensprovider": {
			"legend": {
				"tokenmodifiers": ["declaration", "static", "async", "readonly", "defaultlibrary", "local", "refvalue"],
				"tokentypes": ["class", "enum", "interface", "namespace", "typeparameter", "type", "parameter", "variable", "enummember", "property", "function", "method"]
			},
			"range": true,
			"full": true
		},
		"workspace": {
			"fileoperations": {
				"willrename": {
					"filters": [
						{
							"pattern": {
								"glob": "**/*.{ts,js,vue}"
							}
						}
					]
				}
			},
			"workspacefolders": {
				"changenotifications": true,
				"supported": true
			}
		},
		"completionprovider": {
			"triggercharacters": [".", ":", "<", "\"", "'", "/", "@", "*", " "],
			"resolveprovider": true
		},
		"hoverprovider": true,
		"executecommandprovider": {
			"commands": []
		},
		"documentlinkprovider": {
			"resolveprovider": false
		},
		"referencesprovider": true,
		"definitionprovider": true,
		"documenthighlightprovider": true,
		"textdocumentsync": {
			"didclose": {},
			"save": {},
			"didopen": {},
			"change": {
				"synckind": 2
			}
		},
		"documentformattingprovider": false,
		"documentsymbolprovider": true
	}
}
```

</details>

**Additional context:**

Even when I add a completion capability for completions:
```
	const result: vscode.InitializeResult = {
		capabilities: {
			textDocumentSync: vscode.TextDocumentSyncKind.Incremental,
			completionProvider: {
				triggerCharacters: ['.']
			}
		}
	};
```

When a LSP client sends a request to Volar,
Volar returns this error:
```
Client sends a request to volar:
:: --> LSP-volar textDocument/completion(5): {'position': {'line': 0, 'character': 0}, 'textDocument': {'uri': 'file:///home/predragnikolic/.config/sublime-text/Packages/LSP-volar/index.vue'}}

Volar sends back this:
:: <~~ LSP-volar 5: {'message': 'Unhandled method textDocument/completion', 'code': -32601}
```

So even when Volar has specified capabilities, the server still will not work with other LSP clients (just yet :). Do you have any idea why?

**Comment by johnsoncodehk** [comment]
You need to config which features should enable by `initializationOptions` in language client, because volar split language features to two TS language server to resolved request blocking. (#390)

See https://github.com/yaegassy/coc-volar/blob/master/src/index.ts#L129-L182.

**Comment by predragnikolic** [comment]
Based on the `ServerInitializationOptions` type,
https://github.com/johnsoncodehk/volar/blob/f52f672d4b6667971dbb1d23cc9116083be64563/packages/shared/src/types.ts#L6-L10
the `initializationOptions.typescript.serverPath` string is the only required option that needs to be provided in the `initializationOptions`.

I tried to supply only the `initializationOptions.typescript.serverPath` option, 
and the server starts 
but there are no signs of intellisense.

----

The `initializationOptions.languageFeatures` or `initializationOptions.documentFeatures` are not required based on  the`ServerInitializationOptions` type.

But reguardless I also tried adding both languageFeatures and documentFeatures in the initialization options,
but still no signs of intellisense.

Here is how my initializationOptions looked like:
<details>

```json
// Settings in here override those in "LSP-volar/LSP-volar.sublime-settings"

{
	"command": ["${node_bin}", "/home/predragnikolic/Documents/sandbox/volar/packages/server/out/index.js", "--stdio"],
	"initializationOptions": {
		"typescript": {
			"serverPath": "/home/predragnikolic/.npm-global/lib/node_modules/typescript/lib/typescript.js",
		},
		"languageFeatures": {
			"references": true,
			"definition": true,
			"typeDefinition": true,
			"callHierarchy": false,
			"hover": true,
			"rename": true,
			"renameFileRefactoring": false,
			"signatureHelp": true,
			"codeAction": true,
			"completion": {
			  "defaultTagNameCase": "both",
			  "defaultAttrNameCase": "kebab",
			  "getDocumentNameCasesRequest": false,
			  "getDocumentSelectionRequest": false,
			},
			"documentHighlight": true,
			"documentLink": false,
			"codeLens": false,
			"semanticTokens": false,
			"schemaRequestService": false,
			"diagnostics": true
		},
		"documentFeatures": {
			"selectionRange": true,
			"foldingRange": true,
			"linkedEditingRange": false,
			"documentSymbol": true,
			"documentColor": true,
			"documentFormatting": {
				"defaultPrintWidth": 90
			}
		}
	}
}

```

</details>

----

I am not sure what I am doing wrong.


Even when the LSP client needs to send server capabilities in initialization options (which I think should not be a clients responsibility :) ),
the server would need to respond with a "correct" [InitializeResult](https://microsoft.github.io/language-server-protocol/specifications/specification-3-17/#initializeResult), which it currently doesn't. 
By "correct" I mean to return all capabilities that the server in that current time supports. Currently Volar just returns `textDocumentSync` options.


Additional question:

I saw the code in `coc-volar` and noticed that there are 3 separate language handlers.

Why is there a need to create 3 separate language handlers:
https://github.com/yaegassy/coc-volar/blob/553acb8670cb78f68b6b5ba760dc95d647f98e5d/src/index.ts#L60-L62
with different server capabilities for each?
https://github.com/yaegassy/coc-volar/blob/553acb8670cb78f68b6b5ba760dc95d647f98e5d/src/index.ts#L129-L182
What the difference between 'api', 'doc' and 'html'?

**Reply by johnsoncodehk** [reply]
> Even when the LSP client needs to send server capabilities in initialization options (which I think should not be a clients responsibility :) ),

This is to support distribute features to multiple servers.

> the server would need to respond with a "correct" InitializeResult, which it currently doesn't.

Volar register features after initialized to prevent request trigger before server ready. (https://github.com/johnsoncodehk/volar/blob/master/packages/server/src/registers/registerlanguageFeatures.ts)

> I saw the code in `coc-volar` and noticed that there are 3 separate language handlers.

If all requests are processed on one server, all features will be blocked when the diagnosis is very slow. Auto-complete has a higher priority than diagnosis. Auto-complete should not be blocked, so I assign features to 2 servers (api, doc) according to whether the feature can be blocked. But this is not enough. In the api server, tag-auto-close feature is stateless, it is very fast, but it will be blocked by auto-complete, so I further allocated the stateless features to the html server.

`languageFeatures`, `documentFeatures` options allow the language client to decide how many servers to provide services. If you don't care about sensitivity, you can enable all features in one language server.

**Reply by predragnikolic** [reply]
@johnsoncodehk thanks you for your time!

I didn't knew that Volar supports dynamic registration,
that helped me a lot and it made the server work in Sublime Text.

![image](https://user-images.githubusercontent.com/22029477/130321155-a463029c-fea2-4c71-b648-dc88739a10dc.png)

**Reply by ElMassimo** [reply]
@predragnikolic I'm interested in using Volar in Sublime Text as well, are you planning to submit [LSP Volar](https://github.com/sublimelsp/LSP-volar) to Package Control?

**Reply by predragnikolic** [reply]
Hello @ElMassimo,

In short, I am not planning to push LSP-volar to Package Control,
but I was planning to invite collaborators to the LSP-volar repo. So hopefully other people can publish it to Package Control.

Anyone who is willing to help is more than welcome, 
and I am more than happy to help with getting started. (just ping me :) )
